src/fsharp/FSharp.Core/iseq.fs

```diff
@@ -0,0 +1,1345 @@
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *) })
+
+        [<CompiledName "ForAll2">]
+        let inline forall2 predicate (source1:ISeq<'T>) (source2:ISeq<'U>) : bool =
+            source1.Fold (fun pipeIdx ->
+                upcast { new FolderWithPostProcessing<'T,bool,IEnumerator<'U>>(true,source2.GetEnumerator()) with
+                    override this.ProcessNext value =
+                        if this.State.MoveNext() then
+                            if not (predicate value this.State.Current) then
+                                this.Result <- false
+                                (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        else
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ = ()
+                    override this.OnDispose () = this.State.Dispose () })
+
+        [<CompiledName "Filter">]
+        let inline filter<'T> (f:'T->bool) (source:ISeq<'T>) : ISeq<'T> =
+            source.PushTransform { new ITransformFactory<'T,'T> with
+                override __.Compose _ _ next =
+                    upcast { new Transform<'T,'V,NoValue>(next,Unchecked.defaultof<NoValue>) with
+                        override __.ProcessNext input =
+                            if f input then TailCall.avoid (next.ProcessNext input)
+                            else false } }
+
+        [<CompiledName "Map">]
+        let inline map<'T,'U> (f:'T->'U) (source:ISeq<'T>) : ISeq<'U> =
+            source.PushTransform { new ITransformFactory<'T,'U> with
+                override __.Compose _ _ next =
+                    upcast { new Transform<'T,'V,NoValue>(next,Unchecked.defaultof<NoValue>) with
+                        override __.ProcessNext input =
+                            TailCall.avoid (next.ProcessNext (f input)) } }
+
+        [<CompiledName "MapIndexed">]
+        let inline mapi f (source:ISeq<_>) =
+            source.PushTransform { new ITransformFactory<'T,'U> with
+                override __.Compose _ _ next =
+                    upcast { new Transform<'T,'V,int>(next, -1) with
+                        override this.ProcessNext (input:'T) : bool =
+                            this.State <- this.State  + 1
+                            TailCall.avoid (next.ProcessNext (f this.State input)) } }
+
+        [<CompiledName "Map2">]
+        let inline map2<'T,'U,'V> (map:'T->'U->'V) (source1:ISeq<'T>) (source2:ISeq<'U>) : ISeq<'V> =
+            source1.PushTransform { new ITransformFactory<'T,'V> with
+                override __.Compose outOfBand pipeIdx (next:Activity<'V,'W>) =
+                    upcast { new TransformWithPostProcessing<'T,'W, IEnumerator<'U>>(next, (source2.GetEnumerator ())) with
```

Comment:
Again I'm wondering about how robust the ``Dispose`` on the second enumerator is if early exit occurs because the outer iterator is disposed or because the processing function raises an exception